### PR TITLE
Curp proto/refactor/error

### PR DESCRIPTION
**Pull request by k7Xm9pQr**

### CODE REVIEW

File: src/curp-command.proto [hunk 1]
```diff
@@ -88,8 +138,8 @@ message FetchReadStateResponse {
         repeated ProposeId ids = 1;
     }
     oneof read_state {
-        IdSet ids = 1;
-        uint64 commit_index = 2;
+      IdSet ids = 1;
+      uint64 commit_index = 2;
```
Comment (themanforfree) [inline src/curp-command.proto]: Should we unify indentation length?

File: src/curp-command.proto [hunk 2]
```diff
@@ -103,34 +153,24 @@ message ShutdownResponse {
...
+      Update = 3;
+      Promote = 4;
     }
     message ConfChange {
-        ConfChangeType change_type = 1;
-        uint64 node_id = 2;
-        repeated string address = 3;
+      ConfChangeType change_type = 1;
+      uint64 node_id = 2;
+      repeated string address = 3;
```
Comment (themanforfree) [inline src/curp-command.proto]: Same as above

Comment (k7Xm9pQr) [reply src/curp-command.proto]: Done

File: src/curp-command.proto (hunk 1 above)
Comment (k7Xm9pQr) [reply src/curp-command.proto]: Done

**Review by themanforfree — APPROVED** [review]

File: src/curp-command.proto [hunk 3]
```diff
@@ -37,18 +30,75 @@ message CmdResult {
...
+        google.protobuf.Empty Duplicated = 2;
+        // When the client id in a proposal id is expired.
+        google.protobuf.Empty ExpiredClientId = 3;
+        // When the conf change config is invalid.
+        google.protobuf.Empty InvalidConfig = 4;
+        // When the conf change proposal remove a node which is not exists.
+        google.protobuf.Empty NodeNotExists = 5;
+        // When the conf change proposal add a node which is exists.
+        google.protobuf.Empty NodeAlreadyExists = 6;
+        // When the conf change proposal promote a lazy learner.
```
Comment (Phoenix500526) [inline src/curp-command.proto:65]: What does a lazy learner mean?

File: src/curp-command.proto [hunk 4]
```diff
@@ -144,13 +184,11 @@ message PublishResponse {
...
 
 service Protocol {
-    rpc Propose (ProposeRequest) returns (ProposeResponse);
-    rpc WaitSynced (WaitSyncedRequest)
-            returns (WaitSyncedResponse);
-    rpc ProposeConfChange (ProposeConfChangeRequest) returns (ProposeConfChangeResponse);
+    rpc Propose(ProposeRequest) returns (ProposeResponse);
+    rpc WaitSynced(WaitSyncedRequest) returns (WaitSyncedResponse);
+    rpc ProposeConfChange(ProposeConfChangeRequest) returns (ProposeConfChangeResponse);
     rpc Publish (PublishRequest) returns (PublishResponse);
```
Comment (Phoenix500526) [inline src/curp-command.proto:190]: Please unify the code style with other rpc services. It should be `rpc Publish(PublishRequest) returns (PublishResponse);`, without a space between `Publish` and `(PublishRequest)`

File: src/curp-command.proto [hunk 5]
```diff
@@ -37,18 +30,75 @@ message CmdResult {
...
+}
+
+message ProposeRequest {
+    // The serialized command
+    // The original type is `Command`
+    bytes command = 1;
+    uint64 cluster_version = 2;
 }
 
 message ProposeResponse {
```
Comment (Phoenix500526) [inline src/curp-command.proto:100]: Why should we remove the `leader_id` and `term` in the `ProposeResponse`?

File: src/curp-command.proto [hunk 6]
```diff
@@ -103,34 +153,24 @@ message ShutdownResponse {
...
-        google.protobuf.Empty NodeNotExists = 2;
-        google.protobuf.Empty NodeAlreadyExists = 3;
-        google.protobuf.Empty LearnerNotCatchUp = 4;
-        ProposeError propose = 5;
-        string other = 6;
-    }
-    repeated Member members = 7;
-    optional uint64 leader_id = 8;
-    uint64 term = 9;
+    repeated Member members = 1;
```
Comment (Phoenix500526) [inline src/curp-command.proto:173]: Same as above

File: src/curp-command.proto (hunk 3 above)
Comment (k7Xm9pQr) [reply src/curp-command.proto:65]: A learner did not catch up the leader

File: src/curp-command.proto (hunk 5 above)
Comment (k7Xm9pQr) [reply src/curp-command.proto:100]: I'll try to add later in `ProposeRequest`

File: src/curp-command.proto (hunk 6 above)
Comment (k7Xm9pQr) [reply src/curp-command.proto:173]: RedirectError contains the leader id and term, which should not be cared about here.

File: src/curp-command.proto (hunk 4 above)
Comment (k7Xm9pQr) [reply src/curp-command.proto:190]: Fixed in #23 [ref](https://github.com/xline-kv/curp-proto/pull/23/commits/c82a3663f82354397016bccb7fda55c0b10cf244)

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Pull request merged**